contracts/contracts/TallyNonQv.sol

```diff
@@ -0,0 +1,275 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.10;
+
+import { IMACI } from "./interfaces/IMACI.sol";
+import { Hasher } from "./crypto/Hasher.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { IPoll } from "./interfaces/IPoll.sol";
+import { IMessageProcessor } from "./interfaces/IMessageProcessor.sol";
+import { SnarkCommon } from "./crypto/SnarkCommon.sol";
+import { IVerifier } from "./interfaces/IVerifier.sol";
+import { IVkRegistry } from "./interfaces/IVkRegistry.sol";
+import { CommonUtilities } from "./utilities/CommonUtilities.sol";
+
+/// @title TallyNonQv
+/// @notice The TallyNonQv contract is used during votes tallying
+/// and by users to verify the tally results.
+contract TallyNonQv is Ownable, SnarkCommon, CommonUtilities, Hasher {
+  uint256 private constant TREE_ARITY = 5;
+
+  /// @notice The commitment to the tally results. Its initial value is 0, but after
+  /// the tally of each batch is proven on-chain via a zk-SNARK, it should be
+  /// updated to:
+  ///
+  /// hash2(
+  ///   hashLeftRight(merkle root of current results, salt0)
+  ///   hashLeftRight(number of spent voice credits, salt1),
+  /// )
+  ///
+  /// Where each salt is unique and the merkle roots are of arrays of leaves
+  /// TREE_ARITY ** voteOptionTreeDepth long.
+  uint256 public tallyCommitment;
+
+  uint256 public tallyBatchNum;
+
+  // The final commitment to the state and ballot roots
+  uint256 public sbCommitment;
+
+  IVerifier public immutable verifier;
+  IVkRegistry public immutable vkRegistry;
+  IPoll public immutable poll;
+  IMessageProcessor public immutable messageProcessor;
+
+  /// @notice custom errors
+  error ProcessingNotComplete();
+  error InvalidTallyVotesProof();
+  error AllBallotsTallied();
+  error NumSignUpsTooLarge();
+  error BatchStartIndexTooLarge();
+  error TallyBatchSizeTooLarge();
+
+  /// @notice Create a new Tally contract
+  /// @param _verifier The Verifier contract
+  /// @param _vkRegistry The VkRegistry contract
+  /// @param _poll The Poll contract
+  /// @param _mp The MessageProcessor contract
+  constructor(address _verifier, address _vkRegistry, address _poll, address _mp) payable {
```

Comment:
Can we add checks for zero addresses here and why do we need payable?

Reply:
payable is just to save some little gas on deployment.. and zero address checks are not on other contracts either..